scikit-talk/turndynamics.py

```diff
@@ -0,0 +1,286 @@
+import os
+import glob
+import math
+import re
+import datetime
+import grapheme
+import numpy as np
+import pandas as pd
+from collections import Counter
+
+from tqdm.autonotebook import tqdm
+from joblib import Parallel, delayed
+
+def readcorpus (filename, langshort = None, langfull = None):
+  """ returns a formatted language corpus with turn and transition measures
+
+  :param a: filename of the corpus
+  :type a: string
+  :param b: short version of language name, defaults to none
+  :type b: string, optional
+  :param c: full version of language name, defaults to none
+  :type c: string, optional
+
+  :return: formatted dataframe of the language corpus
+  """
+  # convert time strings to the ISO 8601 time format hh:mm:ss.sss
+  def _converttime(text):
+      if pd.isna(text) == True:
+          return pd.NA
+      else:
+          h,m,s = text.split(':')
+          return int(datetime.timedelta(hours=int(h),
+                                        minutes=int(m),
+                                        seconds=float(s)).total_seconds()*1000)
+
+  # number of unique sources
+  def _getsourceindex(source):
+      return n_sources.index(source) + 1
+
+  # talk, laugh, breath, or other conduct classification
+  def _getnature(utterance):
+      if pd.isna(utterance) == True:
+        return pd.NA
+      if utterance == '[laugh]':
+        return 'laugh'
+      if utterance == '[breath]':
+        return 'breath'
+      if utterance in ['[cough]', '[sneeze]', '[nod]', '[blow]', '[sigh]',
+                        '[yawn]', '[sniff]', '[clearsthroat]',
+                        '[lipsmack]', '[inhales]', '[groan]']:
+        return utterance
+      else:
+        return 'talk'
+
+  # count number of characters
+  def _getnchar(utterance):
+      if pd.isna(utterance) == True:
+        return pd.NA
+      else:
+        utterance = Counter(utterance.replace(" ",""))
+        return sum(utterance.values())
+
+  # create a 'window' for each utterance
```

Comment:
This is a first (untested) python port of my R 'windowed transitions' code. As it needs to run on every single row in the db (creating a 10s window and computing these summary measures for it) it's likely to be quite a drag on performance — optimization and modularization needed.